### PR TITLE
[chat] fix gemini strategy

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4646 
#4681 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- [x] DO NOT use `LazyInitContext` in `Gemini` strategy, as it DOES NOT support `from_pretrained` yet.
- [x] Lora is temporarily disabled since it is incompatible with sharded checkpoint. 
- [x] Checkpoint save is slightly modified for conciseness.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Chat/benchmarks/benchmark_opt_lora_dummy.py [hunk 1]
```diff
@@ -76,9 +76,9 @@ def main(args):
     if args.strategy == "ddp":
         strategy = DDPStrategy()
     elif args.strategy == "colossalai_gemini":
-        strategy = GeminiStrategy(placement_policy="cuda", initial_scale=2**5)
+        strategy = GeminiStrategy(placement_policy="auto", initial_scale=2**5)
```
Comment (ver217) [inline applications/Chat/benchmarks/benchmark_opt_lora_dummy.py:79]: This is not equivalence. Bounding `"static"` to `"colossalai_gemini"` and `"auto"` to `"colossalai_gemini_auto"` is better.

File: applications/Chat/coati/trainer/strategies/base.py [hunk 2]
```diff
@@ -111,7 +111,7 @@ def unwrap_model(model: nn.Module) -> nn.Module:
         return model
 
     def save_model(self, model: nn.Module, path: str, only_rank0: bool = True, **kwargs) -> None:
-        self.booster.save_model(model, path, shard=not only_rank0, **kwargs)
+        self.booster.save_model(model, path, shard=False, **kwargs)
```
Comment (ver217) [inline applications/Chat/coati/trainer/strategies/base.py:114]: Why not shard?

File: applications/Chat/coati/trainer/strategies/colossalai.py [hunk 3]
```diff
@@ -160,6 +159,9 @@ def __init__(
         plugin_initializer = lambda: GeminiPlugin(
             chunk_init_device=get_current_device(),
             placement_policy=placement_policy,
+            shard_param_frac=1.0,
```
Comment (ver217) [inline applications/Chat/coati/trainer/strategies/colossalai.py:162]: Keep the default value

File: applications/Chat/coati/trainer/strategies/colossalai.py [hunk 4]
```diff
@@ -188,8 +190,9 @@ def setup_distributed(self) -> None:
         colossalai.launch_from_torch({}, seed=self.seed)
 
     def model_init_context(self):
-        return LazyInitContext(default_device=get_current_device())
+        return super().model_init_context()
 
     def unwrap_model(self, model: nn.Module) -> nn.Module:
-        assert isinstance(model, GeminiDDP)
-        return model.module
+        ddp_model = model.unwrap()
```
Comment (ver217) [inline applications/Chat/coati/trainer/strategies/colossalai.py:196]: What's the type of `model` here?

File: applications/Chat/coati/trainer/strategies/ddp.py [hunk 5]
```diff
@@ -98,18 +98,18 @@ def save_pretrained(
...
-        model_path = os.path.join(path, "pytorch_model.bin")
-        self.save_model(model, model_path, only_rank0=only_rank0)
+        # model_path = os.path.join(path, "pytorch_model.bin")
+        self.save_model(model, path, only_rank0=only_rank0)
 
-        def _replace_keys(model_path: str, replace_fn: Callable):
-            state_dict = torch.load(model_path, map_location="cpu")
-            state_dict = {replace_fn(k): v for k, v in state_dict.items()}
-            torch.save(state_dict, model_path)
+        # def _replace_keys(model_path: str, replace_fn: Callable):
```
Comment (ver217) [inline applications/Chat/coati/trainer/strategies/ddp.py:104]: If theses codes are useless, then delete them

**Comment by flybird11111** [comment]
image
All tests have passed.

**Comment by flybird11111** [comment]
image
All examples run success.

File: applications/Chat/benchmarks/benchmark_opt_lora_dummy.py (hunk 1 above)
Comment (flybird11111) [reply applications/Chat/benchmarks/benchmark_opt_lora_dummy.py]: updated

File: applications/Chat/coati/trainer/strategies/base.py (hunk 2 above)
Comment (flybird11111) [reply applications/Chat/coati/trainer/strategies/base.py]: We need to replace the keys in a checkpoint by removing the "model" prefix and saving it as unsharded. Only one file needs to be replaced. To save it in sharded format can be included in the next pull request.

File: applications/Chat/coati/trainer/strategies/colossalai.py (hunk 3 above)
Comment (flybird11111) [reply applications/Chat/coati/trainer/strategies/colossalai.py]: updated

File: applications/Chat/coati/trainer/strategies/ddp.py (hunk 5 above)
Comment (flybird11111) [reply applications/Chat/coati/trainer/strategies/ddp.py]: This is useful for replacing the prefix of keys in a checkpoint.

File: applications/Chat/coati/trainer/strategies/colossalai.py (hunk 4 above)
Comment (flybird11111) [reply applications/Chat/coati/trainer/strategies/colossalai.py]: updated

**Review by ver217 — APPROVED** [review]

**Review by CWHer — APPROVED** [review]

**Pull request merged**